File: prototype/rails/src/test/domain/reports/listing_status_tally_test.rb
require "test_helper"

module Domain
  module Reports
    class ListingStatusTallyTest < ActiveSupport::TestCase
      test "it returns every status in lifecycle order" do
        tally = ListingStatusTally.from({})

        assert_equal %w[draft for_sale sold archived], tally.map(&:status)
      end

      test "it reads the count recorded for a status" do
        tally = ListingStatusTally.from({ "for_sale" => 3 })

        assert_equal 3, tally[1].count
      end

      test "it counts a status with no listings as zero" do
        tally = ListingStatusTally.from({ "for_sale" => 3 })

        assert_equal 0, tally[0].count
      end

      test "it totals every status" do
        tally = ListingStatusTally.from({ "for_sale" => 3, "sold" => 2 })

        assert_equal 5, ListingStatusTally.total(tally)
      end

      test "a seller with no listings totals zero" do
        assert_equal 0, ListingStatusTally.total(ListingStatusTally.from({}))
      end
    end
  end
end

File: app/src/app/Domain/Reports/ListingStatusTallyTest.php
<?php

declare(strict_types=1);

namespace App\Domain\Reports;

use App\Domain\Listings\ListingStatus;

it('returns every status in lifecycle order', function (): void {
    $tally = ListingStatusTally::from([]);

    expect(array_map(fn (ListingStatusCount $row): ListingStatus => $row->status, $tally))
        ->toBe([ListingStatus::Draft, ListingStatus::ForSale, ListingStatus::Sold, ListingStatus::Archived]);
});

it('reads the count recorded for a status', function (): void {
    $tally = ListingStatusTally::from([ListingStatus::ForSale->value => 3]);

    expect($tally[1]->count)->toBe(3);
});

it('counts a status with no listings as zero', function (): void {
    $tally = ListingStatusTally::from([ListingStatus::ForSale->value => 3]);

    expect($tally[0]->count)->toBe(0);
});

it('totals every status', function (): void {
    $tally = ListingStatusTally::from([
        ListingStatus::ForSale->value => 3,
        ListingStatus::Sold->value => 2,
    ]);

    expect(ListingStatusTally::total($tally))->toBe(5);
});
